### PR TITLE
IA-669: Home tracking tab

### PR DESCRIPTION
**Pull request by joshbruhin**
More work to be done on this screen, including hooking up some of the data sources for the various screen elements. The screen is generally functional for browsing and launching tasks - 'Tracking' and 'Measuring' tasks - so thought we might want to merge this in now. I'll continue working on the missing functionality. Just let me know if you'd rather not merge yet.

- Add 'UITabViewController' as initial VC
- Add 'TrackingViewController' as first tab (empty VC for the other 3 tabs)
- Add 'TaskBrowsingViewController' for navigating, browsing and launching tasks (this VC can be used on other screens, such as 1.4)
- Add 'ProgressCircleView'

### CODE REVIEW

File: mPower2/mPower2/Custom Views/ProgressCircleView.swift [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    func updateProgress() {
+        
+        let shortestSide: CGFloat = min(self.frame.size.width, self.frame.size.height)
+        let strokeWidth: CGFloat = 4.0
+        let frame = CGRect(x: strokeWidth/2, y: strokeWidth/2, width: shortestSide - strokeWidth, height: shortestSide - strokeWidth)
+        
+        backgroundShape.frame = frame
+        backgroundShape.position = center
+        backgroundShape.path = UIBezierPath(ovalIn: frame).cgPath
+        backgroundShape.strokeColor = UIColor.royal200.cgColor
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/Custom Views/ProgressCircleView.swift]: This is basically the same thing as the `RSDCountdownDial` and for consistency, should use those colors.
`UIColor.rsd_dialRingBackground`

Which, by the by, is **not** purple in the Figma docs. It's gray.

File: mPower2/mPower2/Custom Views/ProgressCircleView.swift [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        backgroundShape.lineWidth = strokeWidth
+        backgroundShape.fillColor = UIColor.white.cgColor
+        
+        progressShape.frame = frame
+        progressShape.path = backgroundShape.path
+        progressShape.position = backgroundShape.position
+        progressShape.strokeColor = UIColor.royal500.cgColor
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/Custom Views/ProgressCircleView.swift]: `UIColor.rsd_dialRing`

File: mPower2/mPower2/Custom Views/ProgressCircleView.swift [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+import UIKit
+
+class ProgressCircleView: UIView {
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/Custom Views/ProgressCircleView.swift]: Can you please set this up as an @IBDesignable with the progress as !IBInspectable so that it will draw in the storyboard? Thanks.

File: mPower2/mPower2/UIColor+mPowerTheme.swift [hunk 4]
```diff
@@ -86,6 +90,14 @@ extension UIColor {
     
     // MARK: Royal Purple
     
+    //    Very Light Purple (unfilled progress circle):
+    //    Royal 200
+    //    #DBD9D9
+    //    219, 217, 217
+    class var royal200: UIColor {
+        return UIColor(red: 219.0 / 255.0, green: 217.0 / 255.0, blue: 217.0 / 255.0, alpha: 1)
+    }
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/UIColor+mPowerTheme.swift]: See comment above. Where are you getting the very light color? From Stockard? It needs to be consistent throughout the app. Elsewhere, the app uses light grey for progress backgrounds on white.

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    }
+    
+    func shouldShowTabs() -> Bool {
+        guard let taskGroups = taskGroups else {
+            return false
+        }
+        return taskGroups.count > 1
+    }
+    
+    func setupView() {
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: Can you please use a storyboard? That makes it easier to visualize the constraints and tweek things for different iPhone sizes. Thanks

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    
+    // MARK: UICollectionViewDelegate
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: kCollectionCellIdentifier, for: indexPath) as? TaskCollectionViewCell
+        if let selectedTaskGroup = selectedTaskGroup {
+            
+            let task = selectedTaskGroup.tasks[indexPath.row]
+            cell?.image = nil
+            task.imageVendor?.fetchImage(for: CGSize(width: 0.0, height: 0.0)) { (_, img) in
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: Please use `.zero` instead of `CGSize(width: 0.0, height: 0.0)` or better yet use the size from the collectionView layout.

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+        
+        guard let selectedTaskGroup = selectedTaskGroup else {
+            return UIEdgeInsetsMake(0, 0, 0, 0)
+        }
+        
+        // We do this so the cells are centered horizontally
+
+        let totalCellWidth = cellSize.width * CGFloat(selectedTaskGroup.tasks.count)
+        let totalSpacingWidth = minCellSpacing * (CGFloat(selectedTaskGroup.tasks.count) - 1)
+        let inset = (collectionView.frame.size.width - CGFloat(totalCellWidth + totalSpacingWidth)) / 2
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: I would prefer to change the spacing rather than the insets and also do the math for a case where there are 4 or 5 cells on an iPhone SE.

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+
+        return UIEdgeInsetsMake(kTopInset, inset, 0, inset)
+    }
+}
+
+protocol TaskBrowserTabViewDelegate {
+    func taskGroupSelected(identifier: String)
+}
+
+class TaskBrowserTabView: UIView {
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: @IBDesignable

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+    
+    @objc func tabSelected() {
+        if let delegate = delegate,
+            let taskGroupIdentifier = taskGroupIdentifier {
+            delegate.taskGroupSelected(identifier: taskGroupIdentifier)
+        }
+    }
+}
+
+class TaskCollectionViewCell: UICollectionViewCell {
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: @IBDesignable and use the storyboard, please.

File: mPower2/mPower2/View Controllers/TrackingViewController.swift [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    // MARK: Actions
+    @IBAction func actionBarTapped(_ sender: Any) {
+        // TODO: jbruhin 5-1-18 implement
+    }
+}
+
+extension TrackingViewController: RSDTaskViewControllerDelegate {
+
+    // MARK: RSTTaskViewControllerDelegate
+    func taskController(_ taskController: RSDTaskController, didFinishWith reason: RSDTaskFinishReason, error: Error?) {
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/View Controllers/TrackingViewController.swift]: Is tapping on these elements suppose to bring up a new instance of the task? I figured they were just visual elements and not suppose to be interactive.

Comment (joshbruhin) [reply mPower2/mPower2/View Controllers/TrackingViewController.swift]: I'm pretty sure they're supposed to be tappable. I'll ask Woody

Comment (syoung-smallwisdom) [reply mPower2/mPower2/View Controllers/TrackingViewController.swift]: I'm confused. Sorry.

What is this vs. TaskBrowserViewController? I thought that was your tab view? It also has an implementation for the task delegate.

File: mPower2/mPower2.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -75,8 +78,6 @@
 		F8D3E7F02062F9DE00CD351C /* BridgeApp.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = F8D3E7A72062F26D00CD351C /* BridgeApp.framework */; };
 		F8D3E7F12062F9DE00CD351C /* BridgeApp.framework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = F8D3E7A72062F26D00CD351C /* BridgeApp.framework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
 		F8D3E7F72063039200CD351C /* BridgeInfo.plist in Resources */ = {isa = PBXBuildFile; fileRef = F8D3E7F62063039200CD351C /* BridgeInfo.plist */; };
-		F8D3E8022063055500CD351C /* sage-mpower-2.pem in Resources */ = {isa = PBXBuildFile; fileRef = F8D3E8002063055500CD351C /* sage-mpower-2.pem */; };
-		F8D3E8032063055500CD351C /* BridgeInfo-private.plist in Resources */ = {isa = PBXBuildFile; fileRef = F8D3E8012063055500CD351C /* BridgeInfo-private.plist */; };
```
[the remark this answers is not in the document]
Comment (joshbruhin) [reply mPower2/mPower2.xcodeproj/project.pbxproj]: fixed

File: mPower2/mPower2/Custom Views/ProgressCircleView.swift (hunk 3 above)
Comment (joshbruhin) [reply mPower2/mPower2/Custom Views/ProgressCircleView.swift]: done

File: mPower2/mPower2/Custom Views/ProgressCircleView.swift (hunk 1 above)
Comment (joshbruhin) [reply mPower2/mPower2/Custom Views/ProgressCircleView.swift]: done

File: mPower2/mPower2/Custom Views/ProgressCircleView.swift (hunk 2 above)
Comment (joshbruhin) [reply mPower2/mPower2/Custom Views/ProgressCircleView.swift]: done

File: mPower2/mPower2/UIColor+mPowerTheme.swift (hunk 4 above)
Comment (joshbruhin) [reply mPower2/mPower2/UIColor+mPowerTheme.swift]: fixed

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift (hunk 5 above)
Comment (joshbruhin) [reply mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: All the content here - including the 'Tracking' and 'Measuring' tabs - is generated dynamically based on the model. The view controller is assigned an array of RSDTaskGroups. A tab is created for each one of the task groups and the content for each task group is derived from the tasks associated with the group. So...I'm not sure what would go into a storyboard. Without the model, which I can't assign in the storyboard, there would be no content.

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift (hunk 6 above)
Comment (joshbruhin) [reply mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: done

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift (hunk 7 above)
Comment (joshbruhin) [reply mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: I added a todo for this

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift (hunk 8 above)
Comment (joshbruhin) [reply mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: done

File: mPower2/mPower2/View Controllers/TaskBrowserViewController.swift (hunk 9 above)
Comment (joshbruhin) [reply mPower2/mPower2/View Controllers/TaskBrowserViewController.swift]: done

File: mPower2/mPower2/View Controllers/TrackingViewController.swift (hunk 10 above)
Comment (joshbruhin) [reply mPower2/mPower2/View Controllers/TrackingViewController.swift]: RSTTaskVC delegate methods removed from this class

File: mPower2/mPower2/UIColor+mPowerTheme.swift [hunk 12]
```diff
@@ -42,6 +42,10 @@ extension UIColor {
         return UIColor.royal500
     }
 
+    @objc open class var veryLightPrimaryTintColor: UIColor {
+        return UIColor.royal400
+    }
+
```
Comment (syoung-smallwisdom) [inline mPower2/mPower2/UIColor+mPowerTheme.swift]: Why is this necessary? It's the same color as `lightPrimaryTintColor` defined below.
